qualtran/bloqs/rotations/hamming_weight_phasing.py

```diff
@@ -86,3 +89,71 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
             (HammingWeightCompute(self.bitsize, adjoint=True), 1),
             (ZPowGate(exponent=self.exponent), self.bitsize.bit_length()),
         }
+
+
+@attrs.frozen
+class HammingWeightPhasingViaPhaseGradient(GateWithRegisters):
+    r"""Applies $Z^{\text{exponent}}$ to every qubit of an input register of size `bitsize`.
+
+    See docstring of `HammingWeightPhasing` for more details about how hamming weight phasing works.
+
+    In this variant of Hamming Weight Phasing, instead of directly synthesizing $O(n.bit_length())$
+    rotations on the hamming weight register, we synthesize the rotations via an addition into the
+    phase gradient register. See reference [1] for more details on this technique.
+
+    Note: For most reasonable values of `bitsize` and `eps`, the naive `HammingWeightPhasing` would
+    have better constant factors than `HammingWeightPhasingViaPhaseGradient`. This is because, in
+    general, the primary advantage of using phase gradient is to reduce the complexity from
+    $O(n * log(1/eps))$ to O(log(1/eps) ** 2) (the phase gradient register is of size
+    $O(log(1/eps))$ and a scaled addition into the target takes $(b_{grad} - 2)(log(1/eps) + 2)$).
+    Therefore, to apply $n$ individual rotations on a target register of size $n$, the complexity is
+    independent of $n$ and is essentially a constant (scales only with log(1/eps)).
+    However, for the actual constant values to be better, value of $n$ needs to be $> log(1/eps)$.
+    In the case of hamming weight phasing, $n$ corresponds to the hamming weight register which itself
+    is $log(\text{bitsize})$. Thus, as `eps` becomes smaller, the required value of $\text{bitsize}$,
+    for the phase gradient version to become more performant, becomes larger.
+
+    Args:
+        bitsize: Size of input register to apply `Z ** exponent` to.
+        exponent: The exponent of `Z ** exponent` to be applied to each qubit in the input register.
+        eps: Accuracy of synthesizing the Z rotations.
+
+    Registers:
+        A single THRU register of size `bitsize`.
+
+    References:
+        [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization]
+        (https://arxiv.org/abs/2007.07391), Appendix A: Addition for controlled rotations
+    """
+
+    bitsize: int
+    exponent: float = 1
+    eps: int = 1e-10
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature.build(x=self.bitsize, phase_grad=self.b_grad)
+
+    @cached_property
+    def b_phase(self) -> int:
+        return int(np.log2(1 / self.eps))
+
+    @cached_property
+    def b_grad(self) -> int:
+        return max(self.bitsize.bit_length(), self.b_phase + int(np.ceil(np.log2(self.b_phase))))
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        x, phase_grad = soqs['x'], soqs['phase_grad']
```

Comment:
Just have x and phase_grad as arguments rather than **soqs?

Reply:
So far, I have been prefering this style because it keeps mypy happy and I don't like the yellow pycharm squiggles in all my classes. I'll change for now but I think this is worth discussing in our weekly sync so everyone follows the same style